File: ruby/lib/data_structures/deque.rb
# frozen_string_literal: true

module DSA
  module DataStructures
    # Double-ended queue supporting operations at both ends.
    # Time: O(1) for all operations
    # Space: O(n)
    class Deque
      Node = Struct.new(:value, :prev_node, :next_node)

      def initialize
        @head = nil
        @tail = nil
        @size = 0
      end

      def push_front(value)
        node = Node.new(value, nil, @head)
        if @head
          @head.prev_node = node
        else
          @tail = node
        end
        @head = node
        @size += 1
        self
      end

      def push_back(value)
        node = Node.new(value, @tail, nil)
        if @tail
          @tail.next_node = node
        else
          @head = node
        end
        @tail = node
        @size += 1
        self
      end

      def pop_front
        return nil unless @head

        value = @head.value
        @head = @head.next_node
        if @head
          @head.prev_node = nil
        else
          @tail = nil
        end
        @size -= 1
        value
      end

      def pop_back
        return nil unless @tail

        value = @tail.value
        @tail = @tail.prev_node
        if @tail
          @tail.next_node = nil
        else
          @head = nil
        end
        @size -= 1
        value
      end

      def peek_front
        @head&.value
      end

      def peek_back
        @tail&.value
      end

      def empty?
        @size.zero?
      end

      def size
        @size
      end

      def clear
        @head = nil
        @tail = nil
        @size = 0
        self
      end
    end
  end
end

File: zig/src/data_structures/deque.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Double-ended queue supporting operations at both ends.
/// Time: O(1) for all operations
/// Space: O(n)
pub fn Deque(comptime T: type) type {
    return struct {
        const Self = @This();
        const Node = struct {
            value: T,
            prev: ?*Node,
            next: ?*Node,
        };

        allocator: Allocator,
        head: ?*Node,
        tail: ?*Node,
        count: usize,

        pub fn init(allocator: Allocator) Self {
            return .{
                .allocator = allocator,
                .head = null,
                .tail = null,
                .count = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
        }

        pub fn pushFront(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .prev = null, .next = self.head };

            if (self.head) |head| {
                head.prev = node;
            } else {
                self.tail = node;
            }
            self.head = node;
            self.count += 1;
        }

        pub fn pushBack(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .prev = self.tail, .next = null };

            if (self.tail) |tail| {
                tail.next = node;
            } else {
                self.head = node;
            }
            self.tail = node;
            self.count += 1;
        }

        pub fn popFront(self: *Self) ?T {
            const head = self.head orelse return null;
            const value = head.value;

            self.head = head.next;
            if (self.head) |new_head| {
                new_head.prev = null;
            } else {
                self.tail = null;
            }

            self.allocator.destroy(head);
            self.count -= 1;
            return value;
        }

        pub fn popBack(self: *Self) ?T {
            const tail = self.tail orelse return null;
            const value = tail.value;

            self.tail = tail.prev;
            if (self.tail) |new_tail| {
                new_tail.next = null;
            } else {
                self.head = null;
            }

            self.allocator.destroy(tail);
            self.count -= 1;
            return value;
        }

        pub fn peekFront(self: *const Self) ?T {
            return if (self.head) |head| head.value else null;
        }

        pub fn peekBack(self: *const Self) ?T {
            return if (self.tail) |tail| tail.value else null;
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.count == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.count;
        }
    };
}

test "Deque operations" {
    const allocator = std.testing.allocator;
    var deque = Deque(i32).init(allocator);
    defer deque.deinit();

    try deque.pushFront(2);
    try deque.pushFront(1);
    try deque.pushBack(3);

    try std.testing.expectEqual(@as(?i32, 1), deque.peekFront());
    try std.testing.expectEqual(@as(?i32, 3), deque.peekBack());
}
